### ISSUE TITLE
Commission: Webcam only from Active/Speaking Users | Reducing Framerate of inactive users

### ISSUE THREAD

**Issue by krim404** [bigbluebutton/bigbluebutton#9392]
is it possible to only display the webcam of users paricipating in the conference? 

We already have the speaking-indicator, this could behave the same way, but with the video-streams, to reduce the load on the client.

Best would be to only show <1FPS for inactive users and full stream for people currently speaking.

**Comment by iSamof** [comment]
I believe this has been mentioned in some other threads but it really will have a huge positive impact on resource efficiency and of course helping increase capacity. So much of BBB capacity usage comes down to webcams.

**Comment by Sea444** [comment]
How should this work? At the moment every participant can chaos a webcam to show in full screen mode. So this webcam should be normal fps I think. So BBB-server should keep on sending out every stream in full resolution/fps and client decides which webcam is shown in full resoultion/fps and which not? So bandwith would be the same but cpu usage is less.
Or should BBB-Server send out only the webcam of talking person (and maybe last 2-3) in full resoultion/fps and the others wit h less resolution/fps?

Would this be a setting fot the server, for a room or for a user?

Concept isn't clear.

**Comment by iSamof** [comment]
> Concept isn't clear.

I guess several colleagues need to comment so we clear up the concept. My thinking (conceptually and not yet practically) is this:

- Each client only send out the webcam video if the user is speaking otherwise it could send a control signal to indicate the client is still connected and is just silent

- the server retransmits the same control code to all other clients to indicate: "hold" last frame for that particular silent user until further notice

As an alternative to holding the last frame of the silent user, the server clients could capture an initial still and transmit to all clients. When the control code for "hold" arrives at any client for any other user, the still for that user is taken from the local cache and displayed until further notice.

Given that bandwidths, capacities and client hardware specs are expected to continue growing, any such arrangement should be optional at the server level and then at the moderator level. 

Additionally, this optimisation could be set to on/off by default for meetings of less than X webcams - X to be configurable and then be on/off by default after X+1 webcams are switched on.

This latter config allows for organisations using an internal network and high spec clients to set a higher X, or not use the optimisation at all.

**Comment by wmchris** [comment]
indeed, this could be on multiple levels:

* Client side (receiver): dont show webcam feed of inactive users based on "last voice event" -> reduce CPU on client, no change on transmission or server, but a slow computer will not crash when 20 people activate their webcam anymore
* Client side (sender): dont send webcam feed or reduce quality / FPS until user is active -> lower ... everything.
* Server Side: choose which feeds are relevant for the client, submit max. amount of streams to user (based on preference or specs) -> lower server load and outgoing traffic, lower client load

At the moment it is possible to crash a session by simply joining it and creating 20 webcam feeds, this way everyone with an older cpu will get to 100% cpu usage and timeout.

**Comment by iSamof** [comment]
Currently work is progressing on another massively useful feature: Layout Editor [PR #8729](https://github.com/bigbluebutton/bigbluebutton/pull/8729/commits)

I wonder if the editor could also allow the option for receiving client to only receive the active speaker's webcam stream. This is a little bit of a reach, diluting the layout editor but it may be a stop-gap until some optimisation process can be evolved - which will probably be months away at best.

**Comment by lars-becker** [comment]
Jitsi has a feature which can limit the transmission to the last X speakers: 
https://github.com/jitsi/jitsi-videobridge/blob/master/doc/last-n.md

**Comment by wmchris** [comment]
okay, so it is possible in webrtc. 
Someone interested in implementing this to BBB? We can do it on a commission base. Please contact me.

**Comment by GremL1N** [comment]
Another way is mixing streams. For example when 8 or more participants streaming webcam, media server mixing this to one stream and sent to client one stream instead of 8 streams.
This method use in some systems. I don't know can kurento do this staf.

**Comment by wmchris** [comment]
> Another way is mixing streams. For example when 8 or more participants streaming webcam, media server mixing this to one stream and sent to client one composite stream instead of 8 streams.
> This method use in some systems. I don't know can kurento do this staf.

this would be a very bad idea, as mixing streams would require a high amount of cpu power on server-side.

**Comment by prlanzarin** [comment]
A Last-N lookalike algorithm is a good idea. No promises, but you've been heard.

**Comment by mtsonline** [comment]
I second the last N view option
and I really also would prefer an optional video mixing above x cams in a room (to be configured per room) because 30 cams would take more cpu than mixing 30 cams and streaming one back I guess.

**Comment by wmchris** [comment]
> I second the last N view option
> and I really also would prefer an optional video mixing above x cams in a room (to be configured per room) because 30 cams would take more cpu than mixing 30 cams and streaming one back I guess.

no. rerendering video will take much longer and more cpu than playing a video back. even with gpu accelleration. Just try to view 30 videos and then try to convert one dvd to an avi.

**Comment by mtsonline** [comment]
> no. rerendering video will take much longer and more cpu than playing 
> a video back. even with gpu accelleration. Just try to view 30 videos 
> and then try to convert one dvd to an avi.
>


But one maybe could render the audio in the same process and freeswitch 
would have less cpu usage? Still, we have enough cpu on the server and 
would like to reduce cpu usage and bandwith of the users, which makes a 
lot sense if you have many webcams in a session.

Always adding: this should be optional / configurable.  But it would 
allow to have many webcams in one room, which many users request

**Comment by wmchris** [comment]
> But one maybe could render the audio in the same process and freeswitch would have less cpu usage? Still, we have enough cpu on the server and would like to reduce cpu usage and bandwith of the users, which makes a lot sense if you have many webcams in a session. Always adding: this should be optional / configurable.  But it would allow to have many webcams in one room, which many users request

many servers are not capable of rendering one video in real time.... and you need at least two... so... no. mixing is no option. only muxing or not even sending something.

**Comment by Daedalus3** [comment]
Just out of out of curiosity: how does Zoom do it?

**Comment by wmchris** [comment]
> Just out of out of curiosity: how does Zoom do it?

afaik: reducing bandwith by reducing framerate and resolution on the fly, up to a still image.

**Comment by quizilkend** [comment]
I'm not sure if this belongs here, but Nexcloud announced performance optimizations for Talk, they should also be dynamic by the number of users.
https://nextcloud.com/blog/open-sourcing-talk-back-end-rc-of-talk-9-brings-lots-of-improvements/

Maybe one could get inspiration there.

**Comment by raoulkramer** [comment]
This would be a very good enhancement that will help to include more users on BBB conferences, especially the one with low resource applications such as mobile phones. [ https://github.com/bigbluebutton/bigbluebutton/issues/9695 ] /

It looks like the multiple webcams pull a lot of resources from the device cpu's with limiting the webcams we could make the client-side application experience better. Right now we have a great server setup but client side devices often can't handle it.

Thanks for the great efforts!

**Comment by prlanzarin** [comment]
Please do not spam multiple issues with the same message :-(.

I'm working on it.

**Comment by raoulkramer** [comment]
Hi @prlanzarin i was just trying to help bundle some similar issues together. Sorry to give the wrong impression. Thanks!

**Comment by prlanzarin** [comment]
@raoulkramer No worries. I've merged the issues with this one.
I'll keep you updated when we have something ready for use on this regard.

**Comment by raoulkramer** [comment]
@prlanzarin thanks keep up the good work!

**Comment by mrmattnewell** [comment]
Hi folks - would it be possible to get some idea of progress on this one?  

Our company is heavily invested in a project to integrate BBB, but in our recent testing we've really struggled to have more than a handful of participants sharing webcams. The problem is especially bad for those with a lower spec machine (the BBB server is not under much stress, but we can see that their CPU's are).  We're assuming it's the same issue?

People have become used to the quality of things like Zoom that don't appear to exhibit the same client CPU limitations for multiple webcam streams - so the feedback we are getting is very poor, putting the project at risk.

Is an update such as the one proposed here in the works - should I tell folks to have faith?

@prlanzarin - apologies to single you out, but you appear to be the chap who knows about these things!

**Comment by prlanzarin** [comment]
I'm working on a bunch of stuff related to this issue, yes. Mainly dynamic camera profiles (automatically reducing camera quality based on number of streams) and a last N based approach.
First one should come soon to develop branch, the last N stuff will take a bit longer. I'm only shipping it when I'm confident it's stable enough.

Keep track of this issue. I plan to keep it as a meta issue and split it into multiple ones the following days.

**Comment by anandphulwani** [comment]
> 
> 
> okay, so it is possible in webrtc.
> Someone interested in implementing this to BBB? We can do it on a commission base. Please contact me.

@wmchris I am interested in developing and implementing it together, i have already reduced the frame rate to 6 via mediaconstraints and it has helped a lot. But still a last N feature and reducing framerate and resolution on the fly, up to a still image will definitely help to scale it better.

**Comment by anandphulwani** [comment]
> 
> 
> I'm working on a bunch of stuff related to this issue, yes. Mainly dynamic camera profiles (automatically reducing camera quality based on number of streams) and a last N based approach.
> First one should come soon to develop branch, the last N stuff will take a bit longer. I'm only shipping it when I'm confident it's stable enough.
> 
> Keep track of this issue. I plan to keep it as a meta issue and split it into multiple ones the following days.

@prlanzarin Can you share your developments on the last N stuff, i can help you out in checking stability of the system as i have required infra to test the whole thing out, as of now, we are using OBS with 4 virtual cameras and supplying them the video files as source to test out, using that with multiple machines, it is easy to test.

**Comment by wmchris** [comment]
@anandphulwani sadly i dont have enough ressources available to implement it by myself. we're very low on personnel. Thats why i offered a commission for a working (OS) implementation.

**Comment by matiasilva** [comment]
For that contact bigbluebutton.org/commercial-support/!

**Comment by anandphulwani** [comment]
> 
> 
> @anandphulwani sadly i dont have enough ressources available to implement it by myself. we're very low on personnel. Thats why i offered a commission for a working (OS) implementation.

I am interested but i would definitely need some insights to get it done quickly, otherwise it would just take a lot of time.

**Comment by ninjuan11** [comment]
@wmchris will do it for at least 500 usd. contact details are on my profile

**Comment by wmchris** [comment]
i contacted ninjuan11, but i doubt he is capable of implementing it - no references at all. still looking. Please be aware requesting prepay with western union will sound very fishy.

**Comment by iSamof** [comment]
Thanks for sharing @wmchris ... only scammers ask for prepay via something like western union.

**Comment by anandphulwani** [comment]
> 
> 
> i contacted ninjuan11, but i doubt he is capable of implementing it - no references at all. Please be aware requesting full prepay with western union will sound suspicious.
> 
> So, still looking.

I looked on his profile, is a fresh profile with absolutely no repos and commits, so i was doubtful, i want to give it a try, pay only once it is implemented, please drop me a personal message, so we can talk over this.

**Comment by ryprfpryr** [comment]
> I'm working on a bunch of stuff related to this issue, yes. Mainly dynamic camera profiles (automatically reducing camera quality based on number of streams) and a last N based approach.
> First one should come soon to develop branch, the last N stuff will take a bit longer. I'm only shipping it when I'm confident it's stable enough.
> 
> Keep track of this issue. I plan to keep it as a meta issue and split it into multiple ones the following days.

Hi @prlanzarin any update on your work? 🙏

**Comment by wmchris** [comment]
> please drop me a personal message, so we can talk over this.

afaik there is no PM system on github.

**Comment by prlanzarin** [comment]
An initial set of changes landed on [2.2.22](https://github.com/bigbluebutton/bigbluebutton/releases/tag/v2.2.22).

There's more stuff being done.

**Comment by ryprfpryr** [comment]
Hi @pedrobmarin first we want to thank you for your great work in 2.2.22, we really appreciate it. We have been testing it in depth with several users using different cameraprofiles intervals and bitrate profiles. 

We still can't get more than 20 webcams working simultaneously. At this point the audio starts to get VERY choppy, (with good or bad bandwidth equally). We have made dozens of tests using different settings, server setups, providers, client locations, etc, etc. also with demo.bigbluebutton.com of course.

I know this is a long history issue and a very requested feature, but I wanted to ask you personally if there's any improvements related to this, set for the upcoming releases. Unfortunately this limitation keep our users comparing the performance with Zoom, Google Meet, etc. on which they just CAN use it this way..

I know this is a recurrent topic and probably annoying (at least for me as a BBB operator and fan..), but I think this is the last step to get over these arguments and let BBB conquer the world.. 💪

Hope you can understand this message not just as a request or question, but also to let you know that we (BBB operators and sysadmins) are very sure that you (core developers) will reach this "milestone" one day.

Thank you very much for your hard work.

**Comment by pedrobmarin** [comment]
No problem. But @prlanzarin deserves some of the credit.

**Comment by ryprfpryr** [comment]
@pedrobmarin of course! I missed his tag, I'm sorry 🙏  @prlanzarin

**Comment by prlanzarin** [comment]
> I know this is a long history issue and a very requested feature, but I wanted to ask you personally if there's any improvements related to this, set for the upcoming releases. Unfortunately this limitation keep our users comparing the performance with Zoom, Google Meet, etc. on which they just CAN use it this way (with +20 webcams on)..

We're completely aware of this. Besides developers, we're operators as well. There's a significant technological gap between BBB's WebRTC implementation and Google Meet's. Don't expect that gap to be closed soon, but there's always going to be work on it. 

As I said, there's more work underway. I won't detail what's shipping or not until it has a production ready PR opened. Current short-term roadmap is
  -  Video pagination
  -  Improvements to the reconnect experience
  -  Last N

**Comment by prlanzarin** [comment]
@ryprfpryr Also you mentioned testing that in demo.bigbluebutton.org. Demo **HASN'T** got the dynamic camera profiles enabled. The `cameraQualityThresholds` configuration is **DISABLED BY DEFAULT** in .22 (see the PR where it was introduced). That's because it has to be sized according to server specs. So, no, demo isn't a testing mirror for that specific thing. Be sure to also enable it on your servers.

**Comment by beatwiz** [comment]
@ryprfpryr I wondered why my testing wasnt producing any better results then i realized the dynamic camera profiles had to be enabled on 

```
/usr/share/meteor/bundle/programs/server/assets/app/config/settings.yml

...
# Entry `thresholds` is an array of:
    # - threshold: minimum number of cameras being shared for profile to applied
    #   profile: a camera profile id from the cameraProfiles configuration array
    #            that will be applied to all cameras when threshold is hit
    cameraQualityThresholds:
      enabled: true
...

```

**Comment by ryprfpryr** [comment]
> >There's a significant technological gap between BBB's WebRTC implementation and Google Meet's. Don't expect that gap to be closed soon, but there's always going to be work on it.

Technologically speaking, which are those differences specifically? @prlanzarin 

> As I said, there's more work underway. I won't detail what's shipping or not until it has a production ready PR opened. Current short-term roadmap is
> 
> * Video pagination
> * Improvements to the reconnect experience
> * Last N

Good to know that! Thank you!

> @ryprfpryr I wondered why my testing wasnt producing any better results then i realized the dynamic camera profiles had to be enabled on
> 
> ```
> /usr/share/meteor/bundle/programs/server/assets/app/config/settings.yml
> 
> ...
> # Entry `thresholds` is an array of:
>     # - threshold: minimum number of cameras being shared for profile to applied
>     #   profile: a camera profile id from the cameraProfiles configuration array
>     #            that will be applied to all cameras when threshold is hit
>     cameraQualityThresholds:
>       enabled: true
> ...
> ```

@beatwiz Yes we were making these tests with it enabled. Anyways my point in all this is beyond this last version or this configuration set particularly. I was talking about BBB future in general. Thank you anyways for your suggestion!

**Comment by prlanzarin** [comment]
> Technologically speaking, which are those differences specifically?
  - Proper transport-wide congestion control
  - Simulcast
  - Single peer connection usage
  - VP9
  - Last N (yes, Google does that with tiled view)
  - ICE restarts
  - Better aggressive nomination support,
  - etc, etc and more etc

Those are all things we want to have someday, somehow. All of them require a good amount of work.

**Comment by prlanzarin** [comment]
ref https://github.com/bigbluebutton/bigbluebutton/pull/10293

**Comment by pbdco** [comment]
@prlanzarin works like a charm!! 👏🚀🙌

**Comment by ryprfpryr** [comment]
Hi @prlanzarin this worked fine for us! Thank you!

**Comment by prlanzarin** [comment]
Cool, thanks for testing. There might still be a few tweaks here and there for 10293 specifically.

**Comment by prlanzarin** [comment]
[Another batch (video pagination) in 2.2.23.](https://github.com/bigbluebutton/bigbluebutton/releases/tag/v2.2.23)

**Comment by ryprfpryr** [comment]
Hi @prlanzarin congratulations and thank you for this great impovement. I've just posted some feedback on the dev list. 

https://groups.google.com/forum/#!topic/bigbluebutton-dev/E7ehAuvzXF4 

```
Hi Anton thank you for this release. Pagination is a great improvement, we tested it out and the first impressions are: 

1- It is very important to keep the presentator on a fixed position in order to see it on every pagination.
2- Negotiation still takes too much time on every page change (i know @prlanzarin is working hard in this).

Thank you
```

Thanks

**Comment by prlanzarin** [comment]
Yes, both acknowledged. Currently working on `2`.

**Comment by ryprfpryr** [comment]
I'm know you're always one step further. Thank you again 🙏

**Comment by prlanzarin** [comment]
Final nail in the coffin: https://github.com/bigbluebutton/bigbluebutton/pull/13869

All points from the OP have been addressed in the past 1+1/2 year and most improvements suggested in comments have also been implemented. _Adieu_.

**Comment by pbdco** [comment]
Thank you again @prlanzarin and congratulations for your incredible work! 💪👏👏👏

Have a great year ahead ! 🥂🍾

**Comment by fariss1070** [comment]
Shut

Le mer. 15 déc. 2021 à 19:46, Paulo Lanzarin ***@***.***> a
écrit :
